pkg/kv/kvtest/store.go

```diff
@@ -527,3 +529,155 @@ func testDeleteWhileIterPrefixSingleSequence(t *testing.T, ms MakeStore, sequenc
+	inPref := uniqueKey("prefin")
+	ininPref := uniqueKey("prefinin")
+
+	allPrefs := [][]byte{extPref, inPref, ininPref}
+
+	// delete a subsection of the read prefix
+	testDeleteWhileIterSamePrefixSingleRun(t, ms, allPrefs, inPref, ininPref)
+
+	// scan a subsection of the delete prefix
+	testDeleteWhileIterSamePrefixSingleRun(t, ms, allPrefs, ininPref, inPref)
+
+	// scan and delete the same prefix
+	testDeleteWhileIterSamePrefixSingleRun(t, ms, allPrefs, ininPref, ininPref)
+}
+
+func testDeleteWhileIterSamePrefixSingleRun(t *testing.T, ms MakeStore, prefsToCreate [][]byte, readPref, delPref []byte) {
+	ctx := context.Background()
+	store := ms(t, ctx)
+	defer store.Close()
+
+	// Emptying the KV store from previous entries. This is essential in these tests as
+	// the verification relies on exact keys to exist
+	cleanIter, err := store.Scan(ctx, delPref)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cleanIter.Close()
+
+	for cleanIter.Next() {
+		e := cleanIter.Entry()
+		if cleanIter != nil {
+			err = store.Delete(ctx, e.Key)
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
+	}
+
+	for _, pref := range prefsToCreate {
+		setupSampleData(t, ctx, store, string(pref), 100)
+	}
+
+	// Will be filled by the scan&read routine, to later verify that all values are passed by scan
+	readDone := make(map[string]bool)
+	deleteDone := make(map[string]bool)
+
+	errGrp, errCtx := errgroup.WithContext(context.Background())
+
+	// Scan and read
+	errGrp.Go(func() error {
+		ei, err := store.Scan(errCtx, readPref)
+		if err != nil {
+			return fmt.Errorf("unexpected error from store.Scan (read): %w", err)
+		}
+		defer ei.Close()
+
+		for ei.Next() {
+			e := ei.Entry()
+			if e != nil {
+				readDone[string(e.Key)] = true
+			} else if ei.Err() != nil {
+				return fmt.Errorf("unexpected error during read iteration: %w", err)
+			}
+		}
+		return nil
+	})
+
+	// Scan and delete
+	errGrp.Go(func() error {
+		ei, err := store.Scan(errCtx, delPref)
+		if err != nil {
+			return fmt.Errorf("unexpected error from store.Scan (delete): %w", err)
+		}
+		defer ei.Close()
+
+		for ei.Next() {
+			e := ei.Entry()
+			if e == nil {
+				return fmt.Errorf("unexpected nil entry during deletion interation: %w", ei.Err())
+			}
+			err = store.Delete(errCtx, e.Key)
+			if err != nil {
+				return fmt.Errorf("unexpected delete failure :%w", err)
+			}
+			deleteDone[string(e.Key)] = true
+		}
+		return nil
+	})
+
+	err = errGrp.Wait()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// verify all readDone entries fits readPref
+	for k := range readDone {
+		if strings.Index(k, string(readPref)) != 0 {
+			t.Fatal("unexpected key read for prefix", k, string(readPref))
+		}
+	}
+
+	// verify all delete entries fits delPref
+	for k := range deleteDone {
+		if strings.Index(k, string(delPref)) != 0 {
+			t.Fatal("unexpected key deleted for prefix", k, string(delPref))
+		}
+	}
+
+	// verify all entires that fits delPref are indeed deleted, i.e. no such entry is left
+	// in the store
```

Comment:
Is there an easy way to validate that nothing else was mistakenly deleted?

Reply:
Yes there is :-)
Added code to collect all entries to an entries array, and iterating it to go through all **created** keys. Keys that are not expected to be deleted (not falling under `delPref`) are expected to be reachable via `store.Get`